File: common/interfaces/player.ts
import { GameObject } from './game-object';
import { Challenge } from './challenges';

export interface Avatar {
  id?: number;
  title?: string;
  name: string;
  src: string;
  isSelected?: boolean;
  isTaken?: boolean;
  price: number;
  owned?: boolean;
}

export enum Status {
  Player = "regular-player",
  Admin = "admin",
  Bot = "bot",
  Disconnected = "disconnected",
  PendingDisconnection = "pendingDisconnection",
}

export enum Behavior {
  Sentient = "sentient",
  Aggressive = "aggressive",
  Defensive = "defensive",
}

export interface Player {
  id: string;
  uid: string;
  attributes: Attributes;
  avatar?: Avatar;
  isActive: boolean;
  name: string;
  status: Status;
  postGameStats: PostGameStats;
  inventory: GameObject[];
  position: Position;
  positionHistory: Position[];
  collectedItems?: number[];
  spawnPosition: Position;
  behavior: Behavior;
  assignedChallenge: Challenge;
}
export interface PlayerRecord {
  uid: string;
  attributes: Attributes;
  postGameStats: PostGameStats;
  name: string;
  dropIns: number[];
  dropOuts: number[];
  avatar: Avatar;
  status: Status;
  assignedChallenge: Challenge;
}
export interface Attributes {
  initialHp?: number;
  initialSpeed?: number;
  initialAttack?: number;
  initialDefense?: number;
  totalHp: number;
  currentHp: number;
  speed: number;
  movementPointsLeft: number;
  maxActionPoints: number;
  actionPoints: number;
  attack: number;
  atkDiceMax: number;
  defense: number;
  defDiceMax: number;
  evasion: number;
}

export interface PostGameStats {
  combats: number;
  victories: number;
  draws: number;
  evasions: number;
  defeats: number;
  damageDealt: number;
  damageTaken: number;
  itemsObtained: number;
  tilesVisited: number;
  doorsInteracted: number;
}

export interface Position {
  x: number;
  y: number;
}

File: client-leger/lib/interfaces/player.dart
import 'package:client_leger/interfaces/attributes.dart';
import 'package:client_leger/interfaces/position.dart';
import 'package:client_leger/models/challenge.dart';

class Player {
  String id;
  String? uid;
  Attributes attributes;
  dynamic avatar;
  bool isActive;
  String name;
  dynamic status;
  dynamic postGameStats;
  List<dynamic> inventory;
  Position position;
  List<dynamic> positionHistory;
  List<dynamic> collectedItems;
  Position spawnPosition;
  dynamic behavior;
  Challenge? assignedChallenge;

  Player({
    required this.id,
    this.uid,
    required this.attributes,
    this.avatar,
    required this.isActive,
    required this.name,
    this.status,
    this.postGameStats,
    required this.inventory,
    required this.position,
    required this.positionHistory,
    required this.collectedItems,
    required this.spawnPosition,
    this.behavior,
    this.assignedChallenge
  });

  factory Player.fromJson(Map<String, dynamic> json) {
    return Player(
      id: json['id'] ?? '',
      uid: json['uid'],
      attributes: Attributes.fromJson(
        Map<String, dynamic>.from(json['attributes']),
      ),
      avatar: json['avatar'] ?? {},
      isActive: json['isActive'] ?? false,
      name: json['name'] ?? '',
      status: json['status'] ?? '',
      postGameStats: json['postGameStats'] ?? {},
      inventory: json['inventory'] ?? [],
      position: Position.fromJson(json['position']),
      positionHistory: json['positionHistory'] ?? [],
      collectedItems: json['collectedItems'] ?? [],
      spawnPosition: Position.fromJson(json['spawnPosition']),
      behavior: json['behavior'] ?? '',
            assignedChallenge: json['assignedChallenge'] != null 
          ? Challenge.fromJson(json['assignedChallenge'])
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'uid': uid,
      'attributes': attributes.toJson(),
      'avatar': avatar,
      'isActive': isActive,
      'name': name,
      'status': status,
      'postGameStats': postGameStats,
      'inventory': inventory,
      'position': position.toJson(),
      'positionHistory': positionHistory,
      'collectedItems': collectedItems,
      'spawnPosition': spawnPosition.toJson(),
      'behavior': behavior,
      'assignedChallenge': assignedChallenge?.toJson(),
    };
  }
}
